File: java/majority_element_ii.java
public class Solution {
    public List<Integer> majorityElement(int[] nums) {
        List<Integer> res=new ArrayList<Integer>();
        int n1=0, n2=0, c1=0, c2=0;
        for (int i=0; i<nums.length; i++) {
            if (c1==0 && c2==0) {
                n1=nums[i];
                c1++;
            } else if (c1==0) {
                if (n2==nums[i]) c2++;
                else {
                    n1=nums[i];
                    c1++;
                }
            } else if (c2==0) {
                if (n1==nums[i]) c1++;
                else {
                    n2=nums[i];
                    c2++;
                }    
            } else {
                if (n1==nums[i]) c1++;
                else if (n2==nums[i]) c2++;
                else {
                    c1--;
                    c2--;
                }
            }
        }
        
        c1=c2=0;
        for (int i=0; i<nums.length; i++) {
            if (nums[i]==n1) c1++;
            else if (nums[i]==n2) c2++;
        }
        
        if (c1>nums.length/3) res.add(n1);
        if (c2>nums.length/3) res.add(n2);
        
        return res;
    }
}

File: cpp/majority_element_II.cpp
229. Majority Element II  
Given an integer array of size n, find all elements that appear more than ⌊ n/3 ⌋ times. The algorithm should run in linear time and in O(1) space.

Hint:

How many majority elements could it possibly have?
Do you have a better hint? Suggest it!
Hide Company Tags Zenefits
Hide Tags Array
Hide Similar Problems (E) Majority Element

class Solution {
public:
    vector<int> majorityElement(vector<int>& nums) {
        vector<int> res;
        int n1=0, n2=0, c1=0, c2=0;
        for (auto n:nums) {
            if (n==n1) c1++;
            else if (n==n2) c2++;
            else if (c1==0) {
                n1=n;
                c1++;
            } else if (c2==0) {
                n2=n;
                c2++;
            } else {
                c1--;
                c2--;
            }
        }
        
        c1=c2=0;
        for (auto n:nums) {
            if (n==n1) c1++;
            else if (n==n2) c2++;
        }
        
        if (c1>nums.size()/3) res.push_back(n1);
        if (c2>nums.size()/3) res.push_back(n2);
        return res;
    }
};

